share/wake/lib/system/io.wake

```diff
@@ -15,6 +15,92 @@
 
 package wake
 
+export data PathType =
+    PathTypeRegularFile
+    PathTypeDirectory
+    PathTypeCharacterDevice
+    PathTypeBlockDevice
+    PathTypeFifo
+    PathTypeSymlink
+    PathTypeSocket
+
+export tuple Permission =
+    Read: Boolean
+    Write: Boolean
+    Execute: Boolean
+
+export def perm2bits (perm: Permission) =
+    def b3 =
+        if getPermissionRead perm then
+            4
+        else
+            0
+
+    def b2 =
+        if getPermissionWrite perm then
+            2
+        else
+            0
+
+    def b1 =
+        if getPermissionExecute perm then
+            1
+        else
+            0
+
+    b1 + b2 + b3
+
+export def bits2perm (bits: Integer) =
+    def b3 = 4 == and bits 4
+    def b2 = 2 == and bits 2
+    def b1 = 1 == and bits 1
+
+    Permission b3 b2 b1
+
+export tuple Mode =
+    User: Permission
+    Group: Permission
+    Other: Permission
+
+export def bits2mode (bits: Integer) =
+    def otherBits = and bits 7
+    def groupBits = and (bits >> 3) 7
+    def userBits = and (bits >> 6) 7
+
+    Mode (bits2perm userBits) (bits2perm groupBits) (bits2perm otherBits)
+
+export def mode2bits (mode: Mode) =
+    def otherBits = perm2bits mode.getModeOther
+    def groupBits = perm2bits mode.getModeGroup << 3
+    def userBits = perm2bits mode.getModeUser << 6
+
+    otherBits + groupBits + userBits
+
+export tuple Stat =
+    Type: PathType
+    Mode: Mode
+    SizeBytes: Integer
```

Comment:
This feels like a good interface; the number of files we'll have to `stat` is much less than the number of `Path`s we produce, so it's likely unnecessary overhead to add to every call.  That said, though, your symlink-vs-standard-file hash conflict does make it seem like we should be taking this into account anyway (and directories created with different modes can have very noticeable differences when trying to write files within them), and if we're `stat`ing everything to be able to differentiate the hashes then we might want to go the `getPathType`/`*Mode`/`*SizeBytes` route to keep the conceptually-related info centralized.

Reply:
Yeah I think in an ideal system we would be hashing the mode and type and as long as we're doing that we should also have the size info available (hashing the size is a nop since hash functions of content are already size sensitive)